### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by thenotsoft**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: src/RequestBodyParser.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    private function getContentType(ServerRequestInterface $request): ?string
+    {
+        $contentType = $request->getHeaderLine(Header::CONTENT_TYPE);
+        if (is_string($contentType) && trim($contentType) !== '') {
+            if (strpos($contentType, ';') !== false) {
+                $contentTypeParts = explode(';', $contentType);
+                return strtolower(trim($contentTypeParts[0]));
+            } else {
```
Comment (romkatsu) [inline src/RequestBodyParser.php]: maybe it's better this way

```php
private function getContentType(ServerRequestInterface $request): ?string
{
    $contentType = $request->getHeaderLine(Header::CONTENT_TYPE);
    if (is_string($contentType) && trim($contentType) !== '') {
        if (str_contains($contentType, ';')) {
            $contentTypeParts = explode(';', $contentType, 2);
            return strtolower(trim($contentTypeParts[0]));
        }

        return strtolower($contentType);
    }
    return null;
}
```

File: src/ParserInterface.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Request\Body;
+
+interface ParserInterface
+{
+    /**
+     * @return array|object|null
```
Comment (romkatsu) [inline src/ParserInterface.php]: It is better to add @throws ParseException

**Review by romkatsu — CHANGES_REQUESTED** [review]

File: src/RequestBodyParser.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+use Yiisoft\Http\Status;
+use Yiisoft\Request\Body\Parser\JsonParser;
+
+final class RequestBodyParser implements MiddlewareInterface
+{
+    private ResponseFactoryInterface $responseFactory;
+    private ContainerInterface $container;
+
+    private array $parsers = [
+        'application/json' => JsonParser::class,
```
Comment (rustamwin) [inline src/RequestBodyParser.php:23]: What about moving to constants?

Comment (thenotsoft) [reply src/RequestBodyParser.php:23]: mime type or $parseres ?

Comment (rustamwin) [reply src/RequestBodyParser.php:23]: mime type

Comment (thenotsoft) [reply src/RequestBodyParser.php:23]: why not, but it would be better if they were in "yiisoft/http" )

Comment (rustamwin) [reply src/RequestBodyParser.php:23]: I'm not against :)

File: composer.json [hunk 4]
```diff
@@ -1,37 +1,49 @@
...
         "wiki": "https://www.yiiframework.com/wiki/",
         "irc": "irc://irc.freenode.net/yii",
-        "source": "https://github.com/yiisoft/_____"
+        "source": "https://github.com/yiisoft/request-body-parsers"
     },
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "php": "^7.4"
+        "php": "^7.4",
```
Comment (xepozz) [inline composer.json]:
```suggestion
        "php": ">=7.4",
```

File: src/Parser/JsonParser.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+final class JsonParser implements ParserInterface
+{
+    private bool $assoc;
+    private int $depth;
+    private int $options;
+
+    public function __construct(
+        bool $assoc = true,
+        int $depth = 512,
+        int $options = JSON_THROW_ON_ERROR | JSON_INVALID_UTF8_IGNORE
```
Comment (xepozz) [inline src/Parser/JsonParser.php:19]: What about to add `JSON_UNESCAPED_UNICODE` to default options?

File: tests/bootstrap.php [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?php
```
Comment (xepozz) [inline tests/bootstrap.php]: There's no need.
See [this](https://github.com/yiisoft/request-body-parsers/blob/fa0aab475d0d81070944c86fa6a7d249591193ab/phpunit.xml.dist#L6) and [this](https://github.com/yiisoft/request-body-parsers/blob/fa0aab475d0d81070944c86fa6a7d249591193ab/phpunit.xml.dist#L10)

File: src/RequestBodyParser.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Http\Header;
+use Yiisoft\Http\Status;
+use Yiisoft\Request\Body\Parser\JsonParser;
+
+final class RequestBodyParser implements MiddlewareInterface
```
Comment (xepozz) [inline src/RequestBodyParser.php:17]:
```suggestion
class RequestBodyParser implements MiddlewareInterface
```

File: src/RequestBodyParser.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+{
+    private ResponseFactoryInterface $responseFactory;
+    private ContainerInterface $container;
+
+    private array $parsers = [
```
Comment (xepozz) [inline src/RequestBodyParser.php:22]:
```suggestion
    protected array $parsers = [
```

File: src/RequestBodyParser.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    public function withoutParsers(string ...$mimeTypes): self
+    {
+        $new = clone $this;
+        if (count($mimeTypes) === 0) {
+            $new->parsers = [];
+            return $new;
+        }
+        foreach ($mimeTypes as $mimeType) {
+            if (trim($mimeType) === '') {
+                throw new \InvalidArgumentException('The mime type cannot be an empty string!');
```
Comment (xepozz) [inline src/RequestBodyParser.php]:
```suggestion
                throw new \InvalidArgumentException('The mime type cannot be an empty string.');
```

File: src/RequestBodyParser.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    public function __construct(ResponseFactoryInterface $responseFactory, ContainerInterface $container)
+    {
+        $this->responseFactory = $responseFactory;
+        $this->container = $container;
+    }
+
+    public function withAddedParser(string $mimeType, string $parserClass): self
+    {
+        if ($mimeType === '') {
+            throw new \InvalidArgumentException('The mime type cannot be an empty string!');
```
Comment (xepozz) [inline src/RequestBodyParser.php]:
```suggestion
        if (trim($mimeType) === '') {
            throw new \InvalidArgumentException('The mime type cannot be an empty string.');
```

File: src/RequestBodyParser.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        if ($parserClass === '') {
+            throw new \InvalidArgumentException('The parser class cannot be an empty string!');
```
Comment (xepozz) [inline src/RequestBodyParser.php]:
```suggestion
            throw new \InvalidArgumentException('The parser class cannot be an empty string.');
```

File: src/RequestBodyParser.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        if (strpos($mimeType, '/') === false) {
+            throw new \InvalidArgumentException('Invalid mime type!');
```
Comment (xepozz) [inline src/RequestBodyParser.php]:
```suggestion
            throw new \InvalidArgumentException('Invalid mime type.');
```

File: src/RequestBodyParser.php [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+                    throw new \RuntimeException(
+                        'Request body media type parser return value must be an array, an object, or null'
+                    );
+                }
+                $request = $request->withParsedBody($parsed);
+            } catch (ParseException $e) {
+                if ($this->badRequestResponse) {
+                    $response = $this->responseFactory->createResponse(Status::BAD_REQUEST);
+                    $response->getBody()->write(Status::TEXTS[Status::BAD_REQUEST]);
+                    return $response;
```
Comment (xepozz) [inline src/RequestBodyParser.php:86]: Would be nice to allow configure the response (feature to the future)

File: src/RequestBodyParser.php [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    ];
+    private bool $badRequestResponse = true;
+
+    public function __construct(ResponseFactoryInterface $responseFactory, ContainerInterface $container)
+    {
+        $this->responseFactory = $responseFactory;
+        $this->container = $container;
+    }
+
+    public function withAddedParser(string $mimeType, string $parserClass): self
```
Comment (xepozz) [inline src/RequestBodyParser.php]: Would be great to make method `validateMimeType` and reuse it in `withoutParsers`

Comment (xepozz) [inline src/RequestBodyParser.php]: What about rename to `withParser`?

File: composer.json (hunk 4 above)
Comment (thenotsoft) [reply composer.json]: why do you need it now?

File: src/RequestBodyParser.php (hunk 7 above)
Comment (thenotsoft) [reply src/RequestBodyParser.php:17]: Do you have a use-case for extending it?

File: src/Parser/JsonParser.php (hunk 5 above)
Comment (roxblnfk) [reply src/Parser/JsonParser.php:19]: This constant not for decoding

File: src/RequestBodyParser.php [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        $new = clone $this;
+        $new->badRequestResponse = false;
+        return $new;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $contentType = $this->getContentType($request);
+
+        if ($contentType && ($parser = $this->getParser($contentType)) !== null) {
```
Comment (roxblnfk) [inline src/RequestBodyParser.php]:
```suggestion
        $parser = $this->getParser($this->getContentType($request));
        if ($parser !== null) {
```

File: src/RequestBodyParser.php [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+        return $handler->handle($request);
+    }
+
+    private function getParser(string $contentType): ?ParserInterface
+    {
+        if (isset($this->parsers[$contentType])) {
+            return $this->container->get($this->parsers[$contentType]);
+        }
+        return null;
```
Comment (roxblnfk) [inline src/RequestBodyParser.php]:
```suggestion
    private function getParser(?string $contentType): ?ParserInterface
    {
        if ($contentType !== null && array_key_exists($contentType, $this->parsers)) {
            return $this->container->get($this->parsers[$contentType]);
        }
        return null;
```

File: src/RequestBodyParser.php [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+    public function withAddedParser(string $mimeType, string $parserClass): self
+    {
+        if (trim($mimeType) === '') {
+            throw new \InvalidArgumentException('The mime type cannot be an empty string.');
+        }
+        if ($parserClass === '') {
+            throw new \InvalidArgumentException('The parser class cannot be an empty string.');
+        }
+        if (strpos($mimeType, '/') === false) {
```
Comment (roxblnfk) [inline src/RequestBodyParser.php]: Are you sure this check is needed here?

File: src/RequestBodyParser.php [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        if (isset($this->parsers[$contentType])) {
+            return $this->container->get($this->parsers[$contentType]);
+        }
+        return null;
+    }
+
+    private function getContentType(ServerRequestInterface $request): ?string
+    {
+        $contentType = $request->getHeaderLine(Header::CONTENT_TYPE);
```
Comment (roxblnfk) [inline src/RequestBodyParser.php:104]: Is it possible that a client sends two or more `Content-Type` headers?

File: src/RequestBodyParser.php (hunk 14 above)
Comment (rustamwin) [reply src/RequestBodyParser.php]: I agree with @xepozz. Because here it replaces the parser.

File: src/RequestBodyParser.php [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+        if (is_string($contentType) && trim($contentType) !== '') {
+            if (str_contains($contentType, ';')) {
+                $contentTypeParts = explode(';', $contentType, 2);
+                return strtolower(trim($contentTypeParts[0]));
+            }
+
+            return strtolower($contentType);
```
Comment (roxblnfk) [inline src/RequestBodyParser.php]: Here the `$contentType` is not trimmed

**Review by roxblnfk — CHANGES_REQUESTED** [review]

File: src/RequestBodyParser.php (hunk 7 above)
Comment (samdark) [reply src/RequestBodyParser.php:17]: I think we'd better leave it final.

File: src/RequestBodyParser.php (hunk 3 above)
Comment (roxblnfk) [reply src/RequestBodyParser.php:23]: [There](https://www.iana.org/assignments/media-types/media-types.xhtml) is full list of mime types.
Are you sure that we need this?

If you need constants, then you can place them in each corresponding parser.
Linke `JsonParser::PREFERRED_CONTENT_TYPE = 'application/json'`

Comment (samdark) [reply src/RequestBodyParser.php:23]: Let's leave it as is for now.

File: composer.json (hunk 4 above)
Comment (roxblnfk) [reply composer.json]: `^7.4` is OK

File: src/RequestBodyParser.php (hunk 17 above)
Comment (samdark) [reply src/RequestBodyParser.php]: [The check is valid](https://tools.ietf.org/html/rfc7231#section-3.1.1.1) but yes, not sure it adds any value here. I'd remove it.

File: src/RequestBodyParser.php (hunk 18 above)
Comment (samdark) [reply src/RequestBodyParser.php:104]: No.

**Comment by samdark** [comment]
Tests should be OK. PHP 8 currently fails because of PHPUnit: https://github.com/sebastianbergmann/phpunit/issues/4325#issuecomment-656678960

File: src/Parser/JsonParser.php [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        $this->convertToAssociativeArray = $convertToAssociativeArray;
+        $this->depth = $depth;
+        $this->options = $options;
+    }
+
+    public function parse(string $rawBody)
+    {
+        try {
+            $result = \json_decode($rawBody, $this->convertToAssociativeArray, $this->depth, $this->options);
+            if (\is_array($result) || \is_object($result)) {
```
Comment (samdark) [inline src/Parser/JsonParser.php:30]: What if it's a string such as

```json
"test"
```
?

File: src/RequestBodyParser.php (hunk 13 above)
Comment (samdark) [reply src/RequestBodyParser.php:86]: I think "bad request" should be enough for all cases. That's not something end user will ever see.

File: src/RequestBodyParser.php (hunk 14 above)
Comment (samdark) [reply src/RequestBodyParser.php]: That check is too simple and too clear to have its own method.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**